src/lib/integration_test_lib/stack_driver_log_engine.ml

```diff
@@ -0,0 +1,513 @@
+    in
+    let _ = parse_log log |> Or_error.ok_exn in
+    ()
+end
+
+module Make (Testnet : Test_intf.Testnet_intf) :
+  Test_intf.Log_engine_intf with type testnet := Testnet.t = struct
+  type subscriptions = {blocks_produced: Subscription.t}
+
+  type t =
+    {testnet_log_filter: string; subscriptions: subscriptions; logger: Logger.t}
+
+  let subscription_list testnet_log_filter : subscriptions Deferred.Or_error.t
+      =
+    (*create one subscription per query*)
+    let open Deferred.Or_error.Let_syntax in
+    let%map blocks_produced =
+      Subscription.create ~name:"blocks_produced"
+        ~filter:(Block_produced_query.filter testnet_log_filter)
+    in
+    {blocks_produced}
+
+  let delete_subscriptions subscriptions =
+    Subscription.delete subscriptions.blocks_produced
+
+  let create ~logger (testnet : Testnet.t) =
+    match%map subscription_list "joyous-occasion" with
+    | Ok subscriptions ->
+        {testnet_log_filter= testnet.testnet_log_filter; subscriptions; logger}
+    | Error e ->
+        failwith (Error.to_string_hum e)
+
+  let cleanup t =
+    match%map delete_subscriptions t.subscriptions with
+    | Ok _ ->
+        None
+    | Error e' ->
+        Logger.fatal t.logger ~module_:__MODULE__ ~location:__LOC__
+          "Error deleting subscriptions: $error"
+          ~metadata:[("error", `String (Error.to_string_hum e'))] ;
+        Some (Error.to_string_hum e')
+
+  let delete t =
+    match%map cleanup t with None -> () | Some err_str -> failwith err_str
+
+  (*TODO: Node status. Should that be a part of a node query instead? or we need a new log that has status info and some node identifier*)
+  let wait_for' :
+         blocks:int
+      -> epoch_reached:int
+      -> timeout:[ `Slots of int
+                 | `Epochs of int
+                 | `Snarked_ledgers_generated of int
+                 | `Milliseconds of int64 ]
+      -> t
+      -> unit Deferred.Or_error.t =
+   fun ~blocks ~epoch_reached ~timeout t ->
+    if blocks = 0 && epoch_reached = 0 && timeout = `Milliseconds 0L then
+      Deferred.Or_error.return ()
+    else
+      let now = Time.now () in
+      let timeout_safety =
+        (*Don't wait for more than an hour in any case*)
+        Time.add now (Time.Span.of_ms (Int64.to_float 3600000L))
```

Comment:
It would be good to add a `TODO` here for better safe timeout guesses for each of the kinds of `timeout` value.

Reply:
We should eventually be moving towards soft/hard timeouts (where we compute a hard timeout from a given soft `~timeout` argument, based one which variant is selected).